Install with "frozen-lockfile" fails when using "injected" dependency

I tried using the `dependenciesMeta.*.injected` feature of PNPM but when I attempt to run `pnpm install --frozen-lockfile` on packages that use `dependenciesMeta.*.injected`, I get the following error even though my lockfile is up to date. 

```
ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up-to-date with package.json
```

I noticed the problem only appears on repos that have set `shared-workspace-lockfile=false` which my repo has. I could not reproduce this error with this flag set to `true`.  Unfortunately I cannot simply change `shared-workspace-lockfile` to `true` in my repo since it would be non-trivial to change (I would need to manually test all 50+ packages in my repo to make sure the new lockfile didn't break anything). 

### pnpm version:

6.22.2

### Code to reproduce the issue:

.pnpm-debug.log: https://gist.github.com/keyworks/bb84f35ff63821370439291ad85f1c7e
Minimal repro example : https://github.com/keyworks/pnpm-injected-test

### Expected behavior:

When running `pnpm install --frozen-lockfile`, it should succeed because the lockfile is up to date.

### Actual behavior:

When running `pnpm install --frozen-lockfile`, I get the following error:
```
jkim@dev-jkim pnpm-injected-test (master) $ pnpm install --frozen-lockfile
Scope: all 3 workspace projects
Already up-to-date
/home/jkim/code/pnpm-injected-test/foo:
 ERR_PNPM_OUTDATED_LOCKFILE  Cannot install with "frozen-lockfile" because pnpm-lock.yaml is not up-to-date with package.json
```

### Additional information:

 - `node -v` prints: v12.21.0
 - Windows, macOS, or Linux?: Linux

--- a/.changeset/cold-humans-end.md
+++ b/.changeset/cold-humans-end.md
@@ -1,0 +1,5 @@
+---
+"pnpm": patch
+---
+
+Install with `--frozen-lockfile` should not fail when the project has injected dependencies and a dedicated lockfile [#4098](https://github.com/pnpm/pnpm/issues/4098).

--- a/.changeset/polite-readers-heal.md
+++ b/.changeset/polite-readers-heal.md
@@ -1,0 +1,5 @@
+---
+"@pnpm/lockfile-file": patch
+---
+
+In a dedicated lockfile the `dependenciesMeta` field should be nested to `'.'` during normalization.

--- a/.changeset/spotty-cows-destroy.md
+++ b/.changeset/spotty-cows-destroy.md
@@ -1,0 +1,5 @@
+---
+"@pnpm/lockfile-file": patch
+---
+
+The `dependenciesMeta` field should be sorted after the dependencies fields.

--- a/packages/lockfile-file/src/read.ts
+++ b/packages/lockfile-file/src/read.ts
@@ -97,6 +97,7 @@ async function _read (
     lockfile.importers = {
       '.': {
         specifiers: lockfile['specifiers'],
+        dependenciesMeta: lockfile['dependenciesMeta'],
       },
     }
     delete lockfile.specifiers

--- a/packages/lockfile-file/src/sortLockfileKeys.ts
+++ b/packages/lockfile-file/src/sortLockfileKeys.ts
@@ -38,8 +38,9 @@ const ROOT_KEYS_ORDER = {
   dependencies: 11,
   optionalDependencies: 12,
   devDependencies: 13,
-  importers: 14,
-  packages: 15,
+  dependenciesMeta: 14,
+  importers: 15,
+  packages: 16,
 }
 
 function compareWithPriority (priority: Record<string, number>, left: string, right: string) {

--- a/packages/lockfile-file/test/read.test.ts
+++ b/packages/lockfile-file/test/read.test.ts
@@ -15,7 +15,17 @@ test('readWantedLockfile()', async () => {
     const lockfile = await readWantedLockfile(path.join('fixtures', '2'), {
       ignoreIncompatible: false,
     })
-    expect(lockfile!.lockfileVersion).toEqual(3)
+    expect(lockfile?.lockfileVersion).toEqual(3)
+    expect(lockfile?.importers).toStrictEqual({
+      '.': {
+        specifiers: {
+          foo: '1',
+        },
+        dependenciesMeta: {
+          foo: { injected: true },
+        },
+      },
+    })
   }
 
   try {

